src/main/java/ch/njol/skript/structures/StructFunction.java

```diff
@@ -134,4 +168,128 @@ public String toString(@Nullable Event event, boolean debug) {
 		return (local ? "local " : "") + "function";
 	}
 
+	public static class FunctionParser {
+
+		/**
+		 * Parses the signature from the given arguments.
+		 *
+		 * @param script  Script file name (<b>might</b> be used for some checks).
+		 * @param name    The name of the function.
+		 * @param args    The parameters of the function.
+		 * @param returns The return type of the function
+		 * @param local   If the signature of function is local.
+		 * @return Parsed signature or null if something went wrong.
+		 * @see Functions#registerSignature(Signature)
+		 */
+		public static @Nullable Signature<?> parse(String script, String name, String args, @Nullable String returns, boolean local) {
+			LinkedHashMap<String, Parameter<?>> parameters = parseParameters(args);
+			if (parameters == null)
+				return null;
+
+			// Parse return type if one exists
+			Class<?> returnType;
+			ClassInfo<?> returnClass;
+
+			if (returns == null) {
+				returnType = null;
+			} else {
+				returnClass = Classes.getClassInfoFromUserInput(returns);
+				PluralResult result = Utils.isPlural(returns);
+
+				if (returnClass == null)
+					returnClass = Classes.getClassInfoFromUserInput(result.updated());
+
+				if (returnClass == null) {
+					Skript.error("Cannot recognise the type '" + returns + "'");
+					return null;
+				}
+
+				if (result.plural()) {
+					returnType = returnClass.getC().arrayType();
+				} else {
+					returnType = returnClass.getC();
+				}
+			}
+
+			return new Signature<>(script, name, parameters, local, returnType, null);
+		}
+
+		private final static Pattern SCRIPT_PARAMETER_PATTERN =
+			Pattern.compile("^\\s*(?<name>[^:(){}\",]+?)\\s*:\\s*(?<type>[a-zA-Z ]+?)\\s*(?:\\s*=\\s*(?<def>.+))?\\s*$");
+
+		private static LinkedHashMap<String, Parameter<?>> parseParameters(String args) {
+			LinkedHashMap<String, Parameter<?>> params = new LinkedHashMap<>();
+
+			boolean caseInsensitive = SkriptConfig.caseInsensitiveVariables.value();
+
+			int j = 0;
+			for (int i = 0; i <= args.length(); i = SkriptParser.next(args, i, ParseContext.DEFAULT)) {
+				if (i == -1) {
+					Skript.error("Invalid text/variables/parentheses in the arguments of this function");
+					return null;
+				}
+
+				if (i == args.length() || args.charAt(i) == ',') {
+					String arg = args.substring(j, i);
+
+					if (args.isEmpty()) // Zero-argument function
+						break;
```

Comment:
this should probably be a few lines higher up for clarity, rather than within the length/comma if